### PR TITLE
Storing metadata as regular data

### PR DESCRIPTION
**Pull request by cw75**
storage tier sends the storage consumption metadata as PUT request to the proxy worker through `metadata_puller`.
proxy metadata thread sends key hotness metadata as PUT request to the proxy worker through `metadata_puller`.

### CODE REVIEW

File: kv_store/lww_kvs/kvs_ebs_server.cpp [hunk 1]
```diff
@@ -1291,19 +1290,24 @@ int main(int argc, char* argv[]) {
...
-      // send the storage consumption update
-      zmq_util::send_string(msg, &cache[monitoring_node.storage_consumption_connect_addr_]);
+      string serialized_req;
+      req.SerializeToString(&serialized_req);
+      // send the storage consumption update to a random proxy worker
+      // just pick the first proxy to contact for now;
+      // this should eventually be round-robin / random
+      string proxy_ip = *(proxy_address.begin());
+      // randomly choose a proxy thread to connect
+      int tid = 1 + rand() % PROXY_THREAD_NUM;
```
[the remark this answers is not in the document]
Comment (cw75) [reply kv_store/lww_kvs/kvs_ebs_server.cpp]: Yes, for now, we are assuming every proxy node has a fixed number of worker thread (probably should be equal to `num_core - 1`)

File: kv_store/lww_kvs/kvs_memory_server.cpp [hunk 2]
```diff
@@ -756,24 +757,24 @@ int main(int argc, char* argv[]) {
...
-      su.set_node_ip(mnode.ip_);
-      su.set_node_type("M");
-      su.set_memory_storage(consumption);
-      su.set_thread_occupancy(float(occupied) / MEMORY_THREAD_NUM);
-      string msg;
-      su.SerializeToString(&msg);
-
-      // send the storage consumption update
-      zmq_util::send_string(msg, &cache[monitoring_node.storage_consumption_connect_addr_]);
+      communication::Request req;
```
[the remark this answers is not in the document]
Comment (cw75) [reply kv_store/lww_kvs/kvs_memory_server.cpp]: So basically every monitoring stats get routed to a proxy worker thread, and the worker thread treat it as a regular `PUT` request and send the updates to the storage tier. The monitoring node periodically query these stats by sending `GET` requests to the proxy worker threads

File: kv_store/lww_kvs/kvs_monitoring.cpp [hunk 3]
```diff
@@ -245,19 +243,96 @@ int main(int argc, char* argv[]) {
...
+          tp->set_key(it->first + "_" + *iter + "_hotness");
+        }
+      }
+
+      string serialized_req;
+      req.SerializeToString(&serialized_req);
+      // send the storage consumption update to a random proxy worker
+      // just pick the first proxy to contact for now;
+      // this should eventually be round-robin / random
+      string proxy_ip = *(proxy_address.begin());
```
[the remark this answers is not in the document]
Comment (cw75) [reply kv_store/lww_kvs/kvs_monitoring.cpp]: Sorry I should have removed the comment that says `// send the storage consumption update to a random proxy worker`. Here the monitoring node is just periodically querying the monitoring stats by sending a `GET` request to the proxy worker thread (The proxy worker thread then route the `GET` request to the storage tier).

[the remark this answers is not in the document]
Comment (vsreekanti) [reply kv_store/lww_kvs/kvs_monitoring.cpp]: Cool, makes sense.

**Pull request merged**